src/cargo/core/resolver/mod.rs

```diff
@@ -244,149 +190,240 @@ struct Context {
-/// context will be passed to the `finished` callback provided.
-fn activate(mut cx: Box<Context>,
+/// If `parent` was activated, this function returns the dependency frame to
+/// iterate through next.
+fn activate(cx: &mut Context,
             registry: &mut Registry,
-            parent: &Rc<Summary>,
-            method: &Method,
-            finished: &mut FnMut(Box<Context>, &mut Registry) -> ResolveResult)
-            -> ResolveResult {
+            parent: Rc<Summary>,
+            method: &Method)
+            -> CargoResult<Option<DepsFrame>> {
     // Dependency graphs are required to be a DAG, so we keep a set of
     // packages we're visiting and bail if we hit a dupe.
-    let id = parent.package_id();
+    let id = parent.package_id().clone();
     if !cx.visited.insert(id.clone()) {
         return Err(human(format!("cyclic package dependency: package `{}` \
                                   depends on itself", id)))
     }
 
     // If we're already activated, then that was easy!
-    if cx.flag_activated(parent, method) {
-        cx.visited.remove(id);
-        return finished(cx, registry)
+    if cx.flag_activated(&parent, method) {
+        cx.visited.remove(&id);
+        return Ok(None);
     }
-    trace!("activating {}", parent.package_id());
+    debug!("activating {}", parent.package_id());
 
-    let deps = try!(cx.build_deps(registry, parent, method));
+    let deps = try!(cx.build_deps(registry, &parent, method));
 
-    activate_deps(cx, registry, parent, deps.iter(), 0, &mut |mut cx, registry| {
-        cx.visited.remove(id);
-        finished(cx, registry)
-    })
+    Ok(Some(DepsFrame{
+        parent: parent,
+        remaining_siblings: RcVecIter::new(deps),
+        id: id,
+    }))
 }
 
-/// Activates the dependencies for a package, one by one in turn.
-///
-/// This function will attempt to activate all possible candidates for each
-/// dependency of the package specified by `parent`. The `deps` iterator
-/// provided is an iterator over all dependencies where each element yielded
-/// informs us what the candidates are for the dependency in question.
+#[derive(Clone)]
+struct RcVecIter<Elem> {
+    vec: Rc<Vec<Elem>>,
+    rest: Range<usize>,
+}
+
+impl<Elem> RcVecIter<Elem> {
+    fn new(vec: Vec<Elem>) -> RcVecIter<Elem> {
+        RcVecIter {
+            rest: 0..vec.len(),
+            vec: Rc::new(vec),
+        }
+    }
+}
+impl<Elem> Iterator for RcVecIter<Elem> where Elem: Clone {
+    type Item = (usize, Elem);
+    fn next(&mut self) -> Option<(usize, Elem)> {
+        self.rest.next().and_then(|i| self.vec.get(i).map(|val| (i, val.clone())))
+    }
+}
+
+#[derive(Clone)]
+struct DepsFrame {
+    parent: Rc<Summary>,
+    remaining_siblings: RcVecIter<DepInfo>,
+    id: PackageId,
+}
+
+struct BacktrackFrame {
+    context_backup: Context,
+    deps_backup: Vec<DepsFrame>,
+    remaining_candidates: RcVecIter<Rc<Summary>>,
```

Comment:
In the interest of cutting down on allocations and reducing the amount of redundantly stored information, I think that this may want to actually just not be stored separately and use perhaps a totally custom iterator type. This is just a filtered list of `all_candidates` below and with `prev_active` already calculated on each iteration of the loop I think it should be possible to just walk among the list of all candidates without storing a pre-filtered list of actual candidates.